### ISSUE TITLE
[ENH] 

### ISSUE THREAD

**Issue by rjungbeck** [cython/cython#4852]
I am using Cython to compile Python applications into native executables (that still require a Python installation  with the referenced libraries). With compiling I mean to convert the Python modules into C files, compiling them and statically linking the resulting objects into one executable. This results in 1 executable for the whole application (and not a shared library per module (which could be hundreds in my case)).

I am using cython_freeze.py to create the main entry point.
But cython_freeze.py can not handle packages

So I implemented a simple workaround in cython_freeze   (by renaming the subpackages). I know that this is not perfect, but my understanding of Cython inner working is not good enough to come up with a full solution (that handles identical named modules and __init__.py).

Maybe someone is interested in implementing a full solution  (I am willing to share my changes of course).

NB: A solution could also be to compile each sub package into it's own shared objects (ie a .so or .dll). This would require cython_freeze.py to create the  DLL or .SO  startup code  (instead of main).

**Comment by da-woods** [comment]
There's a couple of third-party packages that do something similar: https://github.com/smok-serwis/snakehouse and https://github.com/dave-msk/merak. However I think they target making a large bundled .so file rather than an application.

I'm a bit sceptical of the usefulness of trying to build a self-contained application Cython files. The real problem is if you use any third-party modules then you have to bundle them too, and that seems to be quite a hard problem. Personally I'd rather recommend a dedicated solution (e.g. cx_freeze, pyinstaller, nuitka) than try to support it poorly in Cython.

I think modest improvements to the cython_freeze demo script would be welcome. I'd be reluctant to go much further mainly because I'm not personally interested in supporting the full tool that can handle any dependency. Other Cython maintainers may feel differently though (I suspect they don't, but I don't know for sure...).

**Comment by scoder** [comment]
For packaging Python applications (whether its binary modules are built by Cython or not), there are much better tools like pyinstaller or pyrun. See #2011 and #2849.

For wrapping multiple Cython built extension modules into a single, importable extension, it would be nice to improve `cython_freeze` or even add a package compilation mode to `cythonize`. PR would be great.

However, I'm also fine with leaving this to third-party tools, since it's not a very common need and there isn't really a reason why Cython should handle it itself.

**Comment by rjungbeck** [comment]
You are right that this does not reduce the number of library files (which can be more than 11.000 (!!!) for quite simple tools that need matplotlib eg) that need to be distributed. But it protects the applications  source code (which is no longer distributed) and can not be easily be decompiled from the excecutable and the integrity of the application (no one could temper with the installed application source code anymore) . 
I have a build script, that currently unpacks the Python embedded installation into a new directory and pip installs the required libraries into that directory.  I then add the Cython compiled application to it and pack everything into an Inno Setup based installer.
(I first  used py2exe and later PyInstaller before that, but PyInstaller either requires an unpack at each application start (in single file mode) or again the distribution of multiple files. PyInstaller would also  be an additional dependency that needs to be kept in sync with many libraries (for which has its own recipes, where as my needs approach just needs a standard pip install (and sometimes a simple python310._pth file)).

A tool to somehow pack the library files into the application executable, so that there is just a single executable (with no unpacking to disk at each start of course)  would be even better.

**Comment by mattip** [comment]
Have you seen nuitka https://nuitka.net/? It compiles python code into c and bundles it as a single app.
